src/fdconversion.jl

```diff
@@ -0,0 +1,179 @@
+=#
+
+"""converts from Node to Symbolics expression"""
+function _to_symbolics!(a::T, cache::IdDict, variable_map::IdDict) where {T<:FD.Node}
+    tmp = get(cache, a, nothing)
+    if tmp !== nothing
+        return tmp
+    else
+        if FD.arity(a) === 0
+            if FD.is_constant(a)
+                cache[a] = Symbolics.Num(FD.value(a))
+            elseif FD.is_variable(a)
+                tmp = Symbolics.variable(FD.value(a))
+                cache[a] = tmp
+                variable_map[a] = tmp
+            else
+                throw(ErrorException("Node with 0 children was neither constant nor variable. This should never happen."))
+            end
+        else
+            if FD.arity(a) === 1
+                cache[a] = a.node_value(_to_symbolics!(a.children[1], cache, variable_map))
+            else
+                cache[a] = foldl(a.node_value, _to_symbolics!.(a.children, Ref(cache), Ref(variable_map)))
+            end
+        end
+    end
+end
+
+function to_symbolics(a::T, cache::IdDict=IdDict(), variable_map::IdDict=IdDict()) where {T<:FD.Node}
+    _to_symbolics!(a, cache, variable_map)
+    return cache[a], variable_map
+end
+export to_symbolics
+
+function to_symbolics(a::AbstractArray{T}, cache::IdDict=IdDict(), variable_map::IdDict=IdDict()) where {T<:FD.Node}
+    _to_symbolics!.(a, Ref(cache), Ref(variable_map))
+    return map(x -> cache[x], a), variable_map
+end
+
+"""Converts expression x from Symbolics to FastDifferentiation form. Returns a tuple of the **FD** form of the expression and a Dict that maps Symbolics variables to **FD** variables"""
+function to_fd(x::Real, cache=IdDict(), substitutions=IdDict())
+    result = _to_FD!(x, cache, substitutions)
+    syms = collect(filter(x -> SymbolicUtils.issym(x), keys(cache)))
+    sym_map = Dict(zip(syms, map(x -> cache[x], syms)))
+
+    return result, sym_map
+end
+export to_fd
+
+function to_fd(x::AbstractArray{<:Real})
+    cache = IdDict()
+    substitutions = IdDict()
+    result = _to_FD!.(x, Ref(cache), Ref(substitutions))
+    syms = collect(filter(x -> SymbolicUtils.issym(x), keys(cache)))
+    sym_map = Dict(zip(syms, map(x -> cache[x], syms))) #map from symbolics variables to FD variables
+    return result, sym_map
+end
+
+function _to_FD!(sym_node, cache::IdDict, visited::IdDict)
+    # Substitutions are done on a Node graph, not a SymbolicsUtils.Sym graph. As a consequence the values
+    # in substitutions Dict are Node not Sym type. cache has keys (op,args...) where op is generally a function type but sometimes a Sym, 
+    # and args are all Node types.
+
+    @assert !(typeof(sym_node) <: Symbolics.Arr) "Differentiation of expressions involving arrays and array variables is not yet supported."
+    if SymbolicUtils.istree(Symbolics.unwrap(sym_node))
+        @assert !SymbolicUtils.issym(SymbolicUtils.operation(Symbolics.unwrap(sym_node))) "expressions of the form `q(t)` not yet supported."
```

Comment:
Pretty much all use cases would use a `x(t)` symbol, but they act the same in most Jacobians as `x`. Is there a reason why it wouldn't be supported? It would just be a simple conversion.

Reply:
I have support for expressions of the form q(t) but there is a bug in the code that causes these types of expressions to sometimes fail so for now it is not allowed. It is in the queue to be fixed.

When there is an FD term dq(t)/dt = q̇(t) what do you want me to return as the value of q̇(t)? Something like this: D(t)(q)?

Reply:
the value of q̇(t)

Reply:
Did not understand your answer. If somebody does this in Symbolics:
```julia
using Symbolics

@variables t q(t)

fd_jacobian([q],[t])
```
then (when the bug is fixed and FD properly handles forms like q(t)) FD will compute a term ` q̈`. This is represented with a special node type in FD. When ` q̈` is converted from FD to Symbolics form should it be `Differential(t)*Differential(t)(q))?

Reply:
`Differential(t)(q(t))`

it's a little weird but in this case, specifically `@variables t q(t)` binds `q` to `q(t)`.

Reply:
I will need to dynamically construct `q(t)`. Is there a function to do this? I poked around in the Symbolics.jl src and found this https://github.com/JuliaSymbolics/Symbolics.jl/blob/ee3e4917956dee9ee24e0eebc21a9c9bad20ee41/src/variable.jl#L472 but this doesn't look like the right function to create a `q(t)`. If it is could you give me an example of how it should be called?

Reply:
Maybe something like this? 
```julia
function f(q::Symbol,t::Symbol)
       tmp = @variables $q($t)
       return(tmp[1])
       end
```